src/hydrogen/device/rocBLAS_API.cpp

```diff
@@ -9,6 +10,99 @@ namespace hydrogen
 {
 namespace rocblas
 {
+namespace
+{
+/** @brief Manage host-pointer result semantics.
+ *
+ *  If the memory mode on the handle is already set to DEVICE, this is
+ *  just a passthrough. Otherwise, we grab memory from CUB (fallback
+ *  to the HIP stream-aware allocator) and use that.
+ */
+template <typename T>
+class ResultMgr
+{
+public:
+    ResultMgr(rocblas_handle handle, T* result_ptr)
+    : handle_{handle},
+      result_{result_ptr},
+      device_{nullptr}
+    {
+        rocblas_pointer_mode current_ptr_mode;
+        H_CHECK_ROCBLAS(rocblas_get_pointer_mode(handle_, &current_ptr_mode));
+        if (current_ptr_mode == rocblas_pointer_mode_host)
+        {
+            hipStream_t stream;
+            H_CHECK_ROCBLAS(rocblas_get_stream(handle_, &stream));
+#ifdef HYDROGEN_HAVE_CUB
+            H_CHECK_HIP(cub::MemoryPool().DeviceAllocate(
+                            reinterpret_cast<void**>(&device_),
+                            sizeof(T),
+                            stream));
+#else
+            H_CHECK_HIP(hipMallocAsync(
```

Comment:
Probably want to protect that with an ifdef on rocm/hip versions.

Reply:
The point is valid, but we (implicitly at the moment) require ROCm 5.5.0 or newer, which definitely supports this.

Reply:
Cool, good to know.